src/green/builder.rs

```diff
@@ -17,31 +53,44 @@ impl NodeCache {
         I: IntoIterator<Item = GreenElement>,
         I::IntoIter: ExactSizeIterator,
     {
-        let mut node = GreenNode::new(kind, children);
+        let children = children.into_iter();
+        let num_children = children.len();
+        const MAX_CHILDREN: usize = 3;
         // Green nodes are fully immutable, so it's ok to deduplicate them.
         // This is the same optimization that Roslyn does
         // https://github.com/KirillOsenkov/Bliki/wiki/Roslyn-Immutable-Trees
         //
         // For example, all `#[inline]` in this file share the same green node!
         // For `libsyntax/parse/parser.rs`, measurements show that deduping saves
         // 17% of the memory for green nodes!
-        // Future work: make hashing faster by avoiding rehashing of subtrees.
-        if node.children().len() <= 3 {
-            match self.nodes.get(&node) {
-                Some(existing) => node = existing.clone(),
-                None => assert!(self.nodes.insert(node.clone())),
+        if num_children <= MAX_CHILDREN {
+            let collected_children: SmallVec<[GreenElement; MAX_CHILDREN]> = children.collect();
+            let hash = GreenNodeHash::new(kind, collected_children.iter());
+
+            match self.nodes.get(&hash) {
```

Comment:
Hmm so in case of a hash collision, will the partialEq implementation be used to determine if it is a collision because of true identity?

How would I be able to use the raw_entry api if it is nightly only? AFAIK rust analyzer only supports stable

Reply:
Reading up on it myself, yeah a proper partialEq implementation will work. I'll push it tomorrow

Reply:
> How would I be able to use the raw_entry api if it is nightly only?

[Via `hashbrown`](https://docs.rs/hashbrown/0.7.2/hashbrown/struct.HashMap.html#method.raw_entry_mut), if it's actually required (but it's not in this case).